logger/syslogish/server.go

```diff
@@ -0,0 +1,170 @@
+	s.drainMutex.Lock()
+	defer s.drainMutex.Unlock()
+	s.drain = drain
+}
+
+// Listen starts the server's main loop.
+func (s *Server) Listen() {
+	// Should only ever be called once
+	if !s.listening {
+		s.listening = true
+		go s.receive()
+		go s.process()
+		log.Println("syslogish server running")
+	}
+}
+
+func (s *Server) receive() {
+	// Make buffer the same size as the max for a UDP packet
+	buf := make([]byte, 65535)
+	for {
+		n, _, err := s.conn.ReadFrom(buf)
+		if err != nil {
+			log.Fatal("syslogish server read error", err)
+		}
+		message := strings.TrimSuffix(string(buf[:n]), "\n")
+		select {
+		case s.queue <- message:
+		default:
+		}
+	}
+}
+
+func (s *Server) process() {
+	for message := range s.queue {
+		app, err := getAppName(message)
+		if err != nil {
+			log.Println(err)
+			return
+		}
+		// Get a read lock to ensure the storage adapater pointer can't be updated by another
+		// goroutine in the time between we check if it's nil and the time we invoke .Write() upon
+		// it.
+		s.adapterMutex.RLock()
+		// DONT'T defer unlocking... defered statements are executed when the function returns, but
+		// we are inside an infinite loop here.  If we defer, we would never release the lock.
+		// Instead, release it manually below.
+		if s.storageAdapter != nil {
+			s.storageAdapter.Write(app, message)
+			// We don't bother trapping errors here, so failed writes to storage are silent.  This is by
+			// design.  If we sent a log message to STDOUT in response to the failure, deis-logspout
+			// would read it and forward it back to deis-logger, which would fail again to write to
+			// storage and spawn ANOTHER log message.  The effect would be an infinite loop of
+			// unstoreable log messages that would nevertheless fill up journal logs and eventually
+			// overake the disk.
+			//
+			// Treating this as a fatal event would cause the deis-logger unit to restart-- sending
+			// even more log messages to STDOUT.  The overall effect would be the same as described
+			// above with the added disadvantages of flapping.
+			//
+			// But, do not return preemptively.  It's possible the message can still be sent to
+			// the drain.
+		}
+		s.adapterMutex.RUnlock()
+		// Same storay as above for the lock on the drain
+		s.drainMutex.RLock()
+		if s.drain != nil {
+			s.drain.Send(message)
+			// We don't bother trapping errors here.  The rationale is the same as above.
+		}
+		s.drainMutex.RUnlock()
+	}
+}
+
+func getAppName(message string) (string, error) {
+	match := appRegex.FindStringSubmatch(message)
+	if match == nil {
+		return "", fmt.Errorf("Could not find app name in message: %s", message)
+	}
+	return match[1], nil
+}
+
+// ReadLogs returns a specified number of log lines (if available) for a specified app by
+// delegating to the server's underlying storage.Adapter.
+func (s *Server) ReadLogs(app string, lines int) ([]string, error) {
+	if s.storageAdapter == nil {
```

Comment:
After an adapter has been initialized, a new one can be swapped in, but there isn't really a way for the adapter to get nilled out since (unlike in the case of a drain), there is no notion of a nil adapter.

Still, out of an abundance of caution, I agree with you on this.